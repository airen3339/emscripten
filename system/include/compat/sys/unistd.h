#include <unistd.h>
#include <random.h> // for getentropy()
