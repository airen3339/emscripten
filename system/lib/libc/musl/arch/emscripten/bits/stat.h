/*
 * This definition happens to match the x32 and x86-64 stat layouts.
 * Direct compatibility with them isn't essential, but they are good
 * examples to follow. Note that there's no implicit padding, and
 * there's room for future expansion.
 */
struct stat
{
	dev_t st_dev;
	ino_t st_ino;
	nlink_t st_nlink;

	mode_t st_mode;
	uid_t st_uid;
	gid_t st_gid;
	unsigned int __pad0;
	dev_t st_rdev;
	off_t st_size;
	blksize_t st_blksize;
	blkcnt_t st_blocks;

	struct timespec st_atim;
	struct timespec st_mtim;
	struct timespec st_ctim;
	long long __unused[3];
};
