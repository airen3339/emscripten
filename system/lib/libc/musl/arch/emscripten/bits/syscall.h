#define SYS_exit		  1
#define SYS_read		  3
#define SYS_write		  4
#define SYS_open		  5
#define SYS_link		  9
#define SYS_unlink		 10
#define SYS_chdir		 12
#define SYS_mknod		 14
#define SYS_chmod		 15
#define SYS_getpid		 20
#define SYS_pause		 29
#define SYS_access		 33
#define SYS_nice		 34
#define SYS_sync		 36
#define SYS_rename		 38
#define SYS_mkdir		 39
#define SYS_rmdir		 40
#define SYS_dup		 41
#define SYS_pipe		 42
#define SYS_acct		 51
#define SYS_ioctl		 54
#define SYS_setpgid		 57
#define SYS_umask		 60
#define SYS_dup2		 63
#define SYS_getppid		 64
#define SYS_setsid		 66
#define SYS_setrlimit		 75
#define SYS_getrusage		 77
#define SYS_symlink		 83
#define SYS_readlink		 85
#define SYS_munmap		 91
#define SYS_fchmod		 94
#define SYS_getpriority	 96
#define SYS_setpriority	 97
#define SYS_setitimer		104
#define SYS_getitimer		105
#define SYS_wait4		114
#define SYS_setdomainname	121
#define SYS_uname		122
#define SYS_mprotect		125
#define SYS_getpgid		132
#define SYS_fchdir		133
#define SYS__newselect		142
#define SYS_msync		144
#define SYS_getsid		147
#define SYS_fdatasync		148
#define SYS_mlock		150
#define SYS_munlock		151
#define SYS_mlockall		152
#define SYS_munlockall		153
#define SYS_mremap		163
#define SYS_poll		168
#define SYS_rt_sigqueueinfo	178
#define SYS_pread64		180
#define SYS_pwrite64		181
#define SYS_getcwd		183
#define SYS_ugetrlimit		191
#define SYS_mmap2		192
#define SYS_truncate64		193
#define SYS_ftruncate64	194
#define SYS_stat64		195
#define SYS_lstat64		196
#define SYS_fstat64		197
#define SYS_lchown32		198
#define SYS_getuid32		199
#define SYS_getgid32		200
#define SYS_geteuid32		201
#define SYS_getegid32		202
#define SYS_getgroups32	205
#define SYS_fchown32		207
#define SYS_getresuid32	209
#define SYS_getresgid32	211
#define SYS_chown32		212
#define SYS_mincore		218
#define SYS_madvise		219
#define SYS_madvise1		219
#define SYS_getdents64		220
#define SYS_fcntl64		221
#define SYS_exit_group		252
#define SYS_statfs64		268
#define SYS_fstatfs64		269
#define SYS_fadvise64_64	272
#define SYS_openat		295
#define SYS_mkdirat		296
#define SYS_mknodat		297
#define SYS_fchownat		298
#define SYS_fstatat64		300
#define SYS_unlinkat		301
#define SYS_renameat		302
#define SYS_linkat		303
#define SYS_symlinkat		304
#define SYS_readlinkat		305
#define SYS_fchmodat		306
#define SYS_faccessat		307
#define SYS_pselect6		308
#define SYS_utimensat		320
#define SYS_fallocate		324
#define SYS_dup3		330
#define SYS_pipe2		331
#define SYS_preadv		333
#define SYS_pwritev		334
#define SYS_recvmmsg		337
#define SYS_prlimit64		340
#define SYS_sendmmsg		345
#define SYS_socket		359
#define SYS_socketpair		360
#define SYS_bind		361
#define SYS_connect		362
#define SYS_listen		363
#define SYS_accept4		364
#define SYS_getsockopt		365
#define SYS_setsockopt		366
#define SYS_getsockname		367
#define SYS_getpeername		368
#define SYS_sendto		369
#define SYS_sendmsg		370
#define SYS_recvfrom		371
#define SYS_recvmsg		372
#define SYS_shutdown		373
