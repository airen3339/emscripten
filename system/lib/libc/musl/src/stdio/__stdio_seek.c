#include "stdio_impl.h"

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
	off_t ret;
#ifdef SYS__llseek
	off_t off_hi =
#ifndef __EMSCRIPTEN__
		off >> 32;
#else
		0;
#endif
	if (syscall(SYS__llseek, f->fd, off_hi, off, &ret, whence)<0)
		ret = -1;
#else
	ret = syscall(SYS_lseek, f->fd, off, whence);
#endif
	return ret;
}
