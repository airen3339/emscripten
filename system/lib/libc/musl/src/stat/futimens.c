#include <sys/stat.h>

int futimens(int fd, const struct timespec times[2])
{
#if __EMSCRIPTEN__
	#define AT_EMPTY_PATH 0x1000
	return utimensat(fd, 0, times, AT_EMPTY_PATH);
#else
	return utimensat(fd, 0, times, 0);
#endif
}
