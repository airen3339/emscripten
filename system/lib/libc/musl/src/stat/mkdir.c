#include "syscall.h"
#include <fcntl.h>
#include <sys/stat.h>

int mkdir(const char* path, mode_t mode) {
#ifdef SYS_mkdir
  return syscall(SYS_mkdir, path, mode);
#else
  return syscall(SYS_mkdirat, AT_FDCWD, path, mode);
#endif
}
