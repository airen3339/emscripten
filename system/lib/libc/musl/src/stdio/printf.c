#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int iprintf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfiprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vfprintf(FILE *restrict f, const char *restrict fmt, va_list ap);

int __small_printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

