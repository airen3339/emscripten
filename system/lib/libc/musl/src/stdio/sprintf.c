#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int siprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsiprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vsprintf(char *restrict s, const char *restrict fmt, va_list ap);

int __small_sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

