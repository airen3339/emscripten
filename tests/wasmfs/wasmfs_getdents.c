/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <unistd.h>

void print(struct dirent d, int fd) {
  for (;;) {
    int nread = getdents(fd, &d, sizeof(d));
    if (nread == 0) {
      break;
    }
    printf("d.d_name = %s\n", d.d_name);
    printf("d.d_off = %lld\n", d.d_off);
    printf("d.d_reclen = %hu\n", d.d_reclen);
    printf("d.d_type = %hhu\n\n", d.d_type);
  }
}

int main() {
  // Set up test directories.
  assert(mkdir("working", 0777) != -1);
  assert(mkdir("/working/test", 0777) != -1);

  struct dirent d;

  // Try opening the directory that was just created.
  int fd = open("/working", O_RDONLY | O_DIRECTORY);
  printf("------------- Reading from /working Directory -------------\n");
  print(d, fd);

  // Try reading an invalid fd.
  errno = 0;
  getdents(-1, &d, sizeof(d));
  printf("Errno: %s\n", strerror(errno));
  assert(errno == EBADF);

  // Try passing in a size that is too small.
  // Existing file system doesn't catch this error.
  // https://man7.org/linux/man-pages/man2/getdents.2.html#ERRORS
  errno = 0;
  getdents(fd, &d, sizeof(d) - 1);
  printf("Errno: %s\n", strerror(errno));
  assert(errno == EINVAL);

  // Try to read from a file.
  int fileFd = open("/dev/stdout", O_RDONLY);
  getdents(fileFd, &d, sizeof(d));
  printf("Errno: %s\n\n", strerror(errno));
  assert(errno == ENOTDIR);

  // Try opening the root directory and read its contents.
  fd = open("/", O_RDONLY | O_DIRECTORY);

  printf("------------- Reading from root Directory -------------\n");
  print(d, fd);

  // Try opening the dev directory and read its contents.
  fd = open("/dev", O_RDONLY | O_DIRECTORY);
  printf("------------- Reading from /dev Directory -------------\n");
  print(d, fd);
}
