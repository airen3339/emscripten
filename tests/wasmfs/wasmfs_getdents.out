------------- Reading from /working Directory -------------
d.d_name = .
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

d.d_name = ..
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

d.d_name = test
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

Errno: Bad file descriptor
Errno: Invalid argument
Errno: Not a directory

------------- Reading from root Directory -------------
d.d_name = .
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

d.d_name = ..
d.d_off = 280
d.d_reclen = 280
d.d_type = 8

d.d_name = dev
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

d.d_name = working
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

------------- Reading from /dev Directory -------------
d.d_name = .
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

d.d_name = ..
d.d_off = 280
d.d_reclen = 280
d.d_type = 4

d.d_name = stderr
d.d_off = 280
d.d_reclen = 280
d.d_type = 8

d.d_name = stdin
d.d_off = 280
d.d_reclen = 280
d.d_type = 8

d.d_name = stdout
d.d_off = 280
d.d_reclen = 280
d.d_type = 8
